pip/index.py

```diff
@@ -420,7 +424,23 @@ def find_requirement(self, req, upgrade):
 
         Expects req, an InstallRequirement and upgrade, a boolean
         Returns an InstallationCandidate or None
-        May raise DistributionNotFound or BestVersionAlreadyInstalled"""
+        May raise DistributionNotFound or BestVersionAlreadyInstalled
+        """
+        if (req, upgrade) == self._last_find_requirement[0]:
```

Comment:
This whole "cache the last find requirement" thing still bothers me a lot. I don't like adding state like this to the ``PackageFinder``. Since all HTTP access is currently cached (assuming cache control headers) can we just eat the cost of hitting the cache machinery the second time around and just drop this bit of the change. Another PR could then switch this away from raising ``BestVersionAlreadyInstalled`` to returning something that says "the best version is already installed" and ``DistributionNotFound`` returning something to say "we couldn't find anything" (maybe ``None``?). The standard LRU cache decorators could handle this then which feels like a lot cleaner separation of concerns.